### PR TITLE
Fallback to profile lookup if we don't have membership data for a user+room pair

### PR DESCRIPTION
**Pull request by tadzik**
Inspired by https://github.com/Half-Shot/matrix-appservice-discord/blob/develop/src/matrixeventprocessor.ts#L425, which we should probably integrate in MAB to not duplicate this logic everywhere.

### CODE REVIEW

File: src/Main.ts [hunk 1]
```diff
@@ -540,6 +540,10 @@ export class Main {
         return this.stateStorage?.getState(roomId, eventType, stateKey);
     }
 
+    public async getUserProfile(userId: string): Promise<UserProfile|undefined> {
+        return this.botIntent.matrixClient.getUserProfile(userId);
```
[the remark this answers is not in the document]
Comment (tadzik) [reply src/Main.ts]: Fixed now in b922085caa3579cac97b387692bb36ba58e087a7. Saves us a whole bunch of extra complexity :)

File: changelog.d/628.bugfix [hunk 2]
```diff
@@ -0,0 +1 @@
+Make sure we use correct displaynames and avatars on Slack even if we lost track of room state updates
```
Comment (Half-Shot) [inline changelog.d/628.bugfix:1]: Hm, this feels more like developer talk than something users might understand. Would prefer something like: "Improve reliability of Matrix users being bridged with correct displayname and avatar"

File: src/BridgedRoom.ts [hunk 3]
```diff
@@ -486,7 +486,7 @@ export class BridgedRoom {
         const body: ISlackChatMessagePayload = {
             ...matrixToSlackResult,
             as_user: false,
-            username: user.getDisplaynameForRoom(message.room_id) || matrixToSlackResult.username,
+            username: (await user.getDisplaynameForRoom(message.room_id)) || matrixToSlackResult.username,
```
Comment (Half-Shot) [inline src/BridgedRoom.ts:489]: Can this throw horribly, and will we lose the messages?

File: src/BridgedRoom.ts [hunk 4]
```diff
@@ -505,7 +505,7 @@ export class BridgedRoom {
             }
         }
 
-        const avatarUrl = user.getAvatarUrlForRoom(message.room_id);
+        const avatarUrl = await user.getAvatarUrlForRoom(message.room_id);
```
Comment (Half-Shot) [inline src/BridgedRoom.ts:508]: same deal

File: src/Main.ts [hunk 5]
```diff
@@ -544,6 +544,10 @@ export class Main {
         return this.stateStorage?.getState(roomId, eventType, stateKey);
     }
 
+    public async getUserProfile(userId: string): Promise<UserProfile|undefined> {
+        return this.botIntent.getProfileInfo(userId);
```
Comment (Half-Shot) [inline src/Main.ts:548]: Do we not expose botIntent?, this feels like an odd function to have

**Review by Half-Shot — CHANGES_REQUESTED** [review]
A few comments

File: src/Main.ts (hunk 5 above)
Comment (tadzik) [reply src/Main.ts]: Indeed, it needs not exist – though it did make testing a lot easier ;) Removed in bbb79af825dd035d4a4b826389b917a7e2bb106e, and adjusted the tests accordingly.

File: changelog.d/628.bugfix (hunk 2 above)
Comment (tadzik) [reply changelog.d/628.bugfix]: Good point. Fixed in c921200

File: src/BridgedRoom.ts (hunk 3 above)
Comment (tadzik) [reply src/BridgedRoom.ts:489]: Good call. It cannot since 6e1fa00

**Review by Half-Shot — APPROVED** [review]

**Pull request merged**